### PR TITLE
TIP-760: Index Command

### PR DESCRIPTION
**Pull request by anaelChardan**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

It indexes all products (100 by 100) on ES (no delete before)

Here is the output for 1121 products:

```
1121 products to index
Indexing products 0 to 100
Indexing products 100 to 200
Indexing products 200 to 300
Indexing products 300 to 400
Indexing products 400 to 500
Indexing products 500 to 600
Indexing products 600 to 700
Indexing products 700 to 800
Indexing products 800 to 900
Indexing products 900 to 1000
Indexing products 1000 to 1100
Indexing products 1100 to 1121
1121 products indexed
```
[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('pim:product:index')
+            ->addOption(
+                'page-size',
+                false,
+                InputOption::VALUE_OPTIONAL,
+                'If defined define the size of a bulk',
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php]: `Defines the size of a bulk if provided` ?

File: src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        $productRepository = $this->getContainer()->get('pim_catalog.repository.product');
+        $productIndexer = $this->getContainer()->get('pim_catalog.elasticsearch.product_indexer');
+
+        $pageSize = $input->getOption('page-size');
+        $totalElements = $productRepository->countAll();
+        $numberOfPage = ceil($totalElements / $pageSize);
+
+        $output->writeln(sprintf('<info>%s products to index</info>', $totalElements));
+
+        for($currentPage = 1; $currentPage <= $numberOfPage; $currentPage++) {
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php]: space after for

File: src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            $offset = $pageSize * ($currentPage - 1);
+            $output->writeln(
+                sprintf(
+                    'Indexing products %d to %d',
+                    $offset,
+                    ($offset + $pageSize) < $totalElements ? ($offset + $pageSize) : $totalElements
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php]: `($offset + $pageSize) < $totalElements ? ($offset + $pageSize) : $totalElements - 1;`

(or start at 1)

File: src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php (hunk 1 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php]: "Number of product per page" ?

File: src/Pim/Bundle/CatalogBundle/spec/Command/IndexProductCommandSpec.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        $this->getName()->shouldReturn('pim:product:index');
+    }
+
+    function it_is_a_command()
+    {
+        $this->shouldBeAnInstanceOf(ContainerAwareCommand::class);
+    }
+
+    function it_index_all_products(
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/spec/Command/IndexProductCommandSpec.php]: indexes

**Review by grena — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php (hunk 1 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php]: Ok :)

File: src/Pim/Bundle/CatalogBundle/spec/Command/IndexProductCommandSpec.php (hunk 4 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/spec/Command/IndexProductCommandSpec.php]: thx

File: src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        $numberOfPage = ceil($totalElements / $pageSize);
+
+        $output->writeln(sprintf('<info>%s products to index</info>', $totalElements));
+
+        for($currentPage = 1; $currentPage <= $numberOfPage; $currentPage++) {
+            $offset = $pageSize * ($currentPage - 1);
+            $output->writeln(
+                sprintf(
+                    'Indexing products %d to %d',
+                    $offset,
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Command/IndexProductCommand.php]: offset+1

**Review by arnolanglade — APPROVED** [review]

**Pull request closed without merging**